File: qa/appium/crawler/errorsink.py
"""Read the errors the app recorded about itself (CR163).

Replaces an earlier `logcapture.py` that streamed the device log. That approach
was **disproven by measurement**, and the measurement is worth keeping because
it is the kind of thing that otherwise gets rediscovered: a `simctl spawn log
stream` capture of a live Runner process, taken while the app was calling
`debugPrint`, returned **5,685 lines and zero Dart-origin entries** (2026-08-10).
Every apparent "flutter" hit was a CFBundle resource lookup. Dart stdout/stderr
goes to the process's own stdout; `os_log` never sees it, and `flutter run`
only shows it because that command holds the pipe.

So the app records its own errors instead — `mobile/lib/qa/error_sink.dart`,
installed behind `--dart-define=AMI_QA_SEMANTICS=1` — and this reads them back.

Same shape on Android: read the file out of the app's data directory. That is
why the parse is kept separate from the fetch.
"""

from __future__ import annotations

import json
import subprocess
from pathlib import Path

SINK_FILENAME = "ami_qa_errors.jsonl"


class SinkUnavailable(RuntimeError):
    """The sink file could not be located or read.

    A hard failure on purpose. If this were treated as "no errors", every crawl
    against a mis-built app would report a clean bill of health — the exact
    silent-degradation failure this project's conventions forbid. A crawl that
    cannot read the sink has not tested error reporting at all, and must say so.
    """


def ios_sink_path(udid: str, bundle_id: str) -> Path:
    """Resolve `<app data container>/Documents/ami_qa_errors.jsonl`."""
    result = subprocess.run(
        ["xcrun", "simctl", "get_app_container", udid, bundle_id, "data"],
        capture_output=True, text=True, timeout=60,
    )
    if result.returncode != 0:
        raise SinkUnavailable(
            f"could not resolve the app data container for {bundle_id} on {udid}: "
            f"{result.stderr.strip()}. Is the app installed and has it been launched "
            f"at least once?"
        )
    return Path(result.stdout.strip()) / "Documents" / SINK_FILENAME


def read_sink(path: Path) -> list[dict]:
    if not path.exists():
        raise SinkUnavailable(
            f"{path} does not exist. The build under test was almost certainly made "
            f"WITHOUT --dart-define=AMI_QA_SEMANTICS=1, so QaErrorSink never "
            f"installed. Rebuild with scripts/build_qa_ios_sim.sh. Treating this as "
            f"'no errors found' would report a broken run as a clean one."
        )
    return parse_sink(path.read_text(errors="replace"))


def parse_sink(text: str) -> list[dict]:
    """JSONL -> findings. Tolerates a truncated final line: the app may be
    killed mid-write, and losing one record is much better than losing the run."""
    findings: list[dict] = []
    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            record = json.loads(line)
        except json.JSONDecodeError:
            continue
        findings.append(_to_finding(record))
    return findings


# Recognised Flutter error shapes, mapped to a rule id + why it matters. The
# match is on the message the framework produced, so it is stable across our own
# refactors. Unrecognised errors are still reported — as `flutter_error`, with
# the message intact — because an unfamiliar error is more interesting than a
# familiar one, not less.
_SIGNATURES: tuple[tuple[str, str, str, str], ...] = (
    ("overflowed by", "render_overflow", "high",
     "content is being clipped — the classic small-screen / long-translation defect"),
    ("setState() called after dispose()", "setstate_after_dispose", "high",
     "a callback outlived its widget, usually an un-cancelled future or stream"),
    ("Null check operator used on a null value", "null_check_operator", "high",
     "a `!` assertion failed at runtime"),
    ("LateInitializationError", "late_not_initialized", "high",
     "a `late` field was read before it was assigned"),
    ("Failed assertion", "failed_assertion", "high",
     "a framework or app assertion tripped"),
    ("Unable to load asset", "missing_asset", "medium",
     "a referenced asset is not in the build"),
    ("No Material widget found", "missing_ancestor", "medium",
     "a widget was used outside the ancestor it requires"),
    ("A RenderFlex overflowed", "render_overflow", "high",
     "content is being clipped"),
    ("RenderBox was not laid out", "layout_error", "high",
     "a layout constraint violation — usually an unbounded height/width"),
)


def _to_finding(record: dict) -> dict:
    message = str(record.get("message", ""))
    rule, severity, note = "flutter_error", "high", (
        "the app reported a Flutter error. Not in this tool's known-signature "
        "list, which makes it more interesting than a familiar one, not less."
    )
    for needle, rule_id, sev, why in _SIGNATURES:
        if needle.lower() in message.lower():
            rule, severity, note = rule_id, sev, why
            break

    return {
        "rule": rule,
        # `screen` is intentionally absent: the sink has no notion of which
        # screen was up. The crawl path in report.md is how a reviewer locates
        # it, and the ledger fingerprints these on rule+message so the same
        # error is one finding however many screens provoked it.
        "severity": severity,
        "detail": message[:160],
        "note": note,
        "kind": record.get("kind"),
        "library": record.get("library"),
        "sample": (record.get("full") or message)[:400],
        "stack": (record.get("stack") or "")[:600],
        "suggested_category": "mobile",
    }

File: mobile/lib/qa/error_sink.dart
/// CR163 — QA-only error sink: the app records its own failures where an
/// automated crawler can read them.
///
/// Why this exists rather than reading the device log: **Flutter's Dart output
/// does not reach iOS `os_log`.** Measured 2026-08-10 — a `log stream` capture
/// of a running Runner process returned 5,685 lines and not one Dart-origin
/// entry, while the app was actively calling `debugPrint`. Dart stdout/stderr
/// goes to the process's own stdout, which `simctl spawn log stream` never
/// sees. `flutter run` shows those lines because it holds the pipe; a harness
/// that attaches to an already-running app does not.
///
/// So the app writes them down instead. Same mechanism on both platforms, no
/// log-transport differences to reason about, and it survives the crawler
/// crashing.
///
/// **Off unless `--dart-define=AMI_QA_SEMANTICS=1`.** No file is created, no
/// handler is installed, and `runApp` is untouched in a build a user receives.
///
/// This only sees what Flutter *reports*. In a release build Flutter compiles
/// most framework assertions out, which is why the crawler drives a debug
/// build — see `docs/forward_planning/CR163_autonomous_bug_crawler/`.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';

/// Filename inside the app's documents directory. The crawler resolves the
/// container with `xcrun simctl get_app_container <udid> <bundle> data`.
const qaErrorSinkFilename = 'ami_qa_errors.jsonl';

/// Cap so a single error looping every frame cannot fill the disk — the exact
/// scenario this is built to catch (a RenderFlex overflow re-reports on every
/// frame, thousands of times a minute).
const _maxRecords = 500;

class QaErrorSink {
  QaErrorSink._();

  static final QaErrorSink instance = QaErrorSink._();

  File? _file;
  int _written = 0;
  final Set<String> _seen = <String>{};

  /// Install the handlers and truncate any previous run's file. Safe to call
  /// once, from main(), behind the QA flag.
  Future<void> install() async {
    try {
      final dir = await getApplicationDocumentsDirectory();
      _file = File('${dir.path}/$qaErrorSinkFilename');
      await _file!.writeAsString('');
    } catch (_) {
      // No sink means the crawler reports zero errors, which would read as "the
      // app is clean". Say so on stdout so a `flutter run` session shows it;
      // the crawler separately treats a missing file as a hard failure rather
      // than as an empty result.
      debugPrint('QA ERROR SINK UNAVAILABLE — crawl results will be incomplete');
      return;
    }

    final previousOnError = FlutterError.onError;
    FlutterError.onError = (FlutterErrorDetails details) {
      _record(
        kind: 'flutter_error',
        library: details.library ?? 'unknown',
        message: details.exceptionAsString(),
        stack: details.stack?.toString(),
      );
      previousOnError?.call(details);
    };

    // Errors that escape the framework entirely — an unawaited future that
    // throws, a platform-channel failure. FlutterError.onError never sees these.
    PlatformDispatcher.instance.onError = (Object error, StackTrace stack) {
      _record(
        kind: 'uncaught',
        library: 'zone',
        message: error.toString(),
        stack: stack.toString(),
      );
      return false; // false = still report it normally; we are observing, not swallowing
    };
  }

  void _record({
    required String kind,
    required String library,
    required String message,
    String? stack,
  }) {
    final file = _file;
    if (file == null || _written >= _maxRecords) return;

    // Deduplicate on kind+library+first line. The same overflow fires every
    // frame; writing each one buries the second, different defect under
    // thousands of copies of the first.
    final firstLine = message.split('\n').first.trim();
    final key = '$kind|$library|$firstLine';
    if (!_seen.add(key)) return;

    _written++;
    try {
      file.writeAsStringSync(
        '${jsonEncode({
          'kind': kind,
          'library': library,
          'message': firstLine,
          'full': message.length > 2000 ? message.substring(0, 2000) : message,
          'stack': stack == null
              ? null
              : (stack.length > 2000 ? stack.substring(0, 2000) : stack),
        })}\n',
        mode: FileMode.append,
        flush: true, // the crawler may read mid-run, and the app may die next frame
      );
    } catch (_) {
      // Never let the act of recording an error throw a second one.
    }
  }
}
